util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,265 @@
+
+// http  listen ports
+const (
+	workerListenPort = "5003"
+	httpPort         = "5301"
+)
+
+var (
+	// Command arguments for each protocol.This supports templates("{}"),
+	// the value in template will be replaced by value in http request
+	// Iperf command args:
+	// -c <destIP> : connect to destIP(as client)
+	// -f K : report format KBytes/sec
+	// -l 20 : read/write buffer size 20 bytes
+	// -b 1M : bandwidth 1 Mbits/sec
+	// -i 1 : report stats every 1 sec
+	// -t duration : run  <duration> seconds
+	// -u : for udp measurement
+	// -e : enhanced reports, gives more metrics for udp
+	// -s : run in server mode
+	// -P numCls: handle <numCls> number of clients before disconnecting
+	udpClientArguments = []string{"-c", "{destIP}", "-u", "-f", "K", "-l", "20", "-b", "1M", "-e", "-i", "1", "-t", "{duration}"}
+	udpServerArguments = []string{"-s", "-f", "K", "-u", "-e", "-i", "{duration}", "-P", "{numCls}"}
+	tcpServerArguments = []string{"-s", "-f", "K", "-i", "{duration}", "-P", "{numCls}"}
+	tcpClientArguments = []string{"-c", "{destIP}", "-f", "K", "-l", "20", "-b", "1M", "-i", "1", "-t", "{duration}"}
+	// Siege command args:
+	// -d1 : random delay between 0 to 1 sec
+	// -t<duration>S : run test for <duration> seconds
+	// -c1 : one concurrent user
+	httpClientArguments = []string{"http://" + "{destIP}" + ":" + httpPort + "/test", "-d1", "-t" + "{duration}" + "S", "-c1"}
+)
+
+func NewWorker() *Worker {
+	return &Worker{}
+}
+
+// Start starts the worker
+func (w *Worker) Start() {
+	w.stopCh = make(chan struct{})
+	w.listenToController()
+}
+
+func (w *Worker) listenToController() {
+	handlers := handlersMap{
+		"/startTCPServer":  w.StartTCPServer,
+		"/startTCPClient":  w.StartTCPClient,
+		"/startUDPServer":  w.StartUDPServer,
+		"/startUDPClient":  w.StartUDPClient,
+		"/startHTTPServer": w.StartHTTPServer,
+		"/startHTTPClient": w.StartHTTPClient,
+		"/metrics":         w.Metrics,
+	}
+	w.startListening(workerListenPort, handlers)
+}
+
+func (w *Worker) startListening(port string, handlers handlersMap) {
+	for urlPath, handler := range handlers {
+		http.HandleFunc(urlPath, handler)
+	}
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		klog.Fatalf("Failed to start http server on port %v: %v", port, err)
+	}
+}
+
+// Metrics returns the metrics collected
+func (w *Worker) Metrics(rw http.ResponseWriter, request *http.Request) {
+	select {
+	case <-w.stopCh:
+		if w.err != nil {
+			klog.Errorf("Error collecting metrics: %v", w.err)
+			w.sendResponse(rw, http.StatusInternalServerError, status{Message: "metrics collection failed: " + w.err.Error()})
+			return
+		}
+		klog.Info("Metrics collected")
+		var reply metricResponse
+		reply.Result = w.parsedResult
+		timeDifference := w.futureTime.Sub(w.startedAt)
+		reply.WorkerStartTime = "StartedAt: " + w.startedAt.String() + " FutureTime: " + w.futureTime.String() + " diffTime: " + timeDifference.String()
```

Comment:
We wanted the workers to start measuring at almost the same time, especially for the N:1 case. When value of N is high we suspect the time of starting measurement of 1st worker and nth worker could differ by many seconds.

Reply:
So that I understand. And I buy that.

But that is why you pass the timestamp to the worker when it should start it's work. But that is future time.
[And I understand what it is for.]

I'm asking what do you need "startedAt" for.
This is the time at which worker received the request but didn't really do anything useful.
I don't understand what this one is for and you still didn't answer that.